hudi-common/src/main/java/org/apache/hudi/metadata/HoodieMetadataPayload.java

```diff
@@ -620,15 +622,78 @@ public static HoodieRecord<HoodieMetadataPayload> createRecordIndexUpdate(String
    * @param secondaryKey Secondary key of the record
    * @param isDeleted    true if this record is deleted
    */
-  public static HoodieRecord<HoodieMetadataPayload> createSecondaryIndex(String recordKey, String secondaryKey, String partitionPath, Boolean isDeleted) {
-
-    HoodieKey key = new HoodieKey(secondaryKey, partitionPath);
-    HoodieMetadataPayload payload = new HoodieMetadataPayload(secondaryKey, new HoodieSecondaryIndexInfo(recordKey, isDeleted));
+  public static HoodieRecord<HoodieMetadataPayload> createSecondaryIndexRecord(String recordKey, String secondaryKey, String partitionPath, Boolean isDeleted) {
+    // the payload key is in the format of "secondaryKey$primaryKey"
+    HoodieKey key = new HoodieKey(constructSecondaryIndexKey(secondaryKey, recordKey), partitionPath);
+    HoodieMetadataPayload payload = new HoodieMetadataPayload(key.getRecordKey(), new HoodieSecondaryIndexInfo(isDeleted));
     return new HoodieAvroRecord<>(key, payload);
   }
 
-  public String getRecordKeyFromSecondaryIndex() {
-    return secondaryIndexMetadata.getRecordKey();
+  public static String getRecordKeyFromSecondaryIndexKey(String key) {
+    // the payload key is in the format of "secondaryKey$primaryKey"
+    // we need to extract the primary key from the payload key
+    checkState(nonEmpty(key) && key.contains(SECONDARY_INDEX_RECORD_KEY_SEPARATOR), "Invalid key format for secondary index payload: " + key);
+    int delimiterIndex = getSecondaryIndexKeySeparatorPosition(key);
+    return unescapeSpecialChars(key.substring(delimiterIndex + 1));
+  }
+
+  public static String getSecondaryKeyFromSecondaryIndexKey(String key) {
```

Comment:
do these need to be public ? 
Or feel free to move a Utils class and we can write UTs directly